File: src/test/java/me/nasukhov/bot/TukitaLearner/command/AnswerQuestionTest.java
package me.nasukhov.bot.TukitaLearner.command;

import me.nasukhov.TukitaLearner.bot.command.AnswerQuestion;
import me.nasukhov.TukitaLearner.bot.command.AskQuestion;
import me.nasukhov.TukitaLearner.bot.io.Channel;
import me.nasukhov.TukitaLearner.bot.io.Input;
import me.nasukhov.TukitaLearner.bot.io.Output;
import me.nasukhov.TukitaLearner.bot.io.User;
import me.nasukhov.TukitaLearner.study.GroupQuestion;
import me.nasukhov.TukitaLearner.study.ProgressRepository;
import me.nasukhov.TukitaLearner.study.Question;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.Optional;
import java.util.UUID;

import static org.mockito.Mockito.*;

public class AnswerQuestionTest {
    private AnswerQuestion handler;
    private AskQuestion askQuestion;
    private ProgressRepository progressRepository;
    private Output output;

    @BeforeEach
    void setup() {
        output = mock(Output.class);
        progressRepository = mock(ProgressRepository.class);
        askQuestion = mock(AskQuestion.class);

        handler = new AnswerQuestion(progressRepository, askQuestion);
    }

    @Test
    void testHandleUnsupportedCommand() {
        Input input = new Input(
                "Some input",
                new Channel("SomeChannelId"),
                new User("SomeUserId", "SomeUserName")
        );

        handler.handle(input, output);

        verifyNoInteractions(output, progressRepository);
    }

    @Test
    public void testHandleAnswerWithInvalidInput() {
        Input input = new Input(
                "qh answer invalidInput",
                new Channel("SomeChannelId"),
                new User("SomeUserId", "SomeUserName")
        );

        handler.handle(input, output);
        verifyNoInteractions(output, progressRepository);
    }

    @Test
    public void testHandleAnswerWhenUserHasAlreadyAnsweredThatQuestionBefore() {
        UUID questionId = UUID.fromString("9b740f73-2766-4d31-9029-c910759ad41b");

        Input input = new Input(
                "qh answer " + questionId + " 2",
                new Channel("SomeChannelId"),
                new User("SomeUserId", "SomeUserName")
        );

        when(progressRepository.hasReplyInChannel("SomeUserId", "SomeChannelId", questionId)).thenReturn(true);

        handler.handle(input, output);

        verify(progressRepository).hasReplyInChannel("SomeUserId", "SomeChannelId", questionId);
        verifyNoMoreInteractions(progressRepository);
        verifyNoInteractions(output);
    }

    @Test
    public void testHandleAnswerForQuestionThatWasNotAskedInChannel() {
        UUID questionId = UUID.fromString("9b740f73-2766-4d31-9029-c910759ad41b");

        Input input = new Input(
                "qh answer " + questionId + " 2",
                new Channel("SomeChannelId"),
                new User("SomeUserId", "SomeUserName")
        );

        when(progressRepository.hasReplyInChannel("SomeUserId", "SomeChannelId", questionId)).thenReturn(false);
        when(progressRepository.findQuestionInChannel(questionId)).thenReturn(Optional.empty());

        handler.handle(input, output);

        verify(progressRepository).hasReplyInChannel("SomeUserId", "SomeChannelId", questionId);
        verify(progressRepository).findQuestionInChannel(questionId);
        verifyNoMoreInteractions(progressRepository);
        verifyNoInteractions(output);
    }

    @Test
    public void testHandleIncorrectAnswer() {
        UUID questionId = UUID.fromString("9b740f73-2766-4d31-9029-c910759ad41b");

        GroupQuestion groupQuestion = new GroupQuestion(
                questionId,
                new Question(123, "2+5 equals to", "seven", new ArrayList<>(){{
                    add("thirteen");
                    add("seven");
                    add("twenty-one");
                }})
        );

        Input input = new Input(
                "qh answer " + questionId + " 3",
                new Channel("SomeChannelId"),
                new User("SomeUserId", "SomeUserName")
        );

        when(progressRepository.hasReplyInChannel("SomeUserId", "SomeChannelId", questionId)).thenReturn(false);
        when(progressRepository.findQuestionInChannel(questionId)).thenReturn(Optional.of(groupQuestion));

        handler.handle(input, output);

        verify(progressRepository).addUserAnswer(questionId, "SomeUserId", "SomeChannelId", false);
        verify(output).write("SomeUserName, увы, правильный ответ «<spoiler>seven</spoiler>».");
    }

    @Test
    public void testHandleCorrectAnswer() {
        UUID questionId = UUID.fromString("9b740f73-2766-4d31-9029-c910759ad41b");

        GroupQuestion groupQuestion = new GroupQuestion(
                questionId,
                new Question(123, "2+5 equals to", "seven", new ArrayList<>(){{
                    add("thirteen");
                    add("seven");
                    add("twenty-one");
                }})
        );

        Input input = new Input(
                "qh answer " + questionId + " 2",
                new Channel("SomeChannelId"),
                new User("SomeUserId", "SomeUserName")
        );

        when(progressRepository.hasReplyInChannel("SomeUserId", "SomeChannelId", questionId)).thenReturn(false);
        when(progressRepository.findQuestionInChannel(questionId)).thenReturn(Optional.of(groupQuestion));

        handler.handle(input, output);

        verify(progressRepository).addUserAnswer(questionId, "SomeUserId", "SomeChannelId", true);
        verify(output).write("SomeUserName, да, правильный ответ «<spoiler>seven</spoiler>».");
    }

    @Test
    public void testHandleCorrectAnswerInDirectMessages() {
        UUID questionId = UUID.fromString("9b740f73-2766-4d31-9029-c910759ad41b");

        GroupQuestion groupQuestion = new GroupQuestion(
                questionId,
                new Question(123, "2+5 equals to", "seven", new ArrayList<>(){{
                    add("thirteen");
                    add("seven");
                    add("twenty-one");
                }})
        );

        Input input = new Input(
                "qh answer " + questionId + " 2",
                new Channel("SomeChannelId", false),
                new User("SomeUserId", "SomeUserName")
        );

        when(progressRepository.hasReplyInChannel("SomeUserId", "SomeChannelId", questionId)).thenReturn(false);
        when(progressRepository.findQuestionInChannel(questionId)).thenReturn(Optional.of(groupQuestion));

        handler.handle(input, output);

        verify(progressRepository).addUserAnswer(questionId, "SomeUserId", "SomeChannelId", true);
        verify(output).write("SomeUserName, да, правильный ответ «<spoiler>seven</spoiler>».");
        verify(askQuestion).handle(input, output);
    }

    @Test
    public void testHandleIncorrectAnswerInDirectMessages() {
        UUID questionId = UUID.fromString("9b740f73-2766-4d31-9029-c910759ad41b");

        GroupQuestion groupQuestion = new GroupQuestion(
                questionId,
                new Question(123, "2+5 equals to", "seven", new ArrayList<>(){{
                    add("thirteen");
                    add("seven");
                    add("twenty-one");
                }})
        );

        Input input = new Input(
                "qh answer " + questionId + " 3",
                new Channel("SomeChannelId", false),
                new User("SomeUserId", "SomeUserName")
        );

        when(progressRepository.hasReplyInChannel("SomeUserId", "SomeChannelId", questionId)).thenReturn(false);
        when(progressRepository.findQuestionInChannel(questionId)).thenReturn(Optional.of(groupQuestion));

        handler.handle(input, output);

        verify(progressRepository).addUserAnswer(questionId, "SomeUserId", "SomeChannelId", false);
        verify(output).write("SomeUserName, увы, правильный ответ «<spoiler>seven</spoiler>».");
        verify(askQuestion).handle(input, output);
    }
}

File: src/test/kotlin/me/nasukhov/tukitalearner/bot/AnswerQuestionTest.kt
package me.nasukhov.tukitalearner.bot

import me.nasukhov.bot.io.Channel
import me.nasukhov.bot.io.Input
import me.nasukhov.bot.io.NoOutput
import me.nasukhov.bot.io.TextOutput
import me.nasukhov.bot.io.User
import me.nasukhov.tukitalearner.study.Group
import me.nasukhov.tukitalearner.study.GroupQuestion
import me.nasukhov.tukitalearner.study.GroupRepository
import me.nasukhov.tukitalearner.study.ProgressTracker
import me.nasukhov.tukitalearner.study.Question
import me.nasukhov.tukitalearner.study.QuestionRepository
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertInstanceOf
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.transaction.annotation.Transactional
import kotlin.test.assertEquals

@SpringBootTest
@Transactional
internal class AnswerQuestionTest(
    @Autowired private val handler: AnswerQuestion,
    @Autowired private val progressTracker: ProgressTracker,
    @Autowired private val questionRepository: QuestionRepository,
    @Autowired private val groupRepository: GroupRepository,
) {
    private lateinit var group: Group

    @BeforeEach
    fun setup() {
        group = Group("SomeGroup123")
        groupRepository.save(group)
        questionRepository.save(
            Question(
                "2+5 equals to",
                "seven",
                listOf("thirteen", "seven", "twenty-one"),
            ),
        )
    }

    @AfterEach
    fun reset() {
        questionRepository.deleteAll()
        // groupRepository.delete(group)
    }

    @Test
    fun handleCorrect() {
        val channel = Channel(group.id)
        val user = User("SomeId", "SomeName")

        val questionResult = progressTracker.createRandomForGroup(group)
        Assertions.assertTrue(questionResult.isPresent)
        val newGroupQuestion = questionResult.get()
        Assertions.assertFalse(newGroupQuestion.hasAnswerFromUser(user.name))

        val questionId: String? = newGroupQuestion.id.toString()
        val correctAnswerId = getCorrectVariant(newGroupQuestion)

        val input = Input("qh answer $questionId $correctAnswerId", channel, user)

        val output = handler.handle(input)
        assertInstanceOf<TextOutput>(output)
        assertEquals(
            "${user.name}, да, правильный ответ «<spoiler>${newGroupQuestion.viewAnswer()}</spoiler>».",
            output.value,
        )

        assertInstanceOf<NoOutput>(handler.handle(input))
        // TODO check that answer is recorded exactly once
    }

    private fun getCorrectVariant(groupQuestion: GroupQuestion): Int {
        var variantId = 0

        val correctAnswer = groupQuestion.viewAnswer()
        while (groupQuestion.viewVariant(variantId) != correctAnswer) {
            check(variantId++ < 1000) { "Looks like an endless loop" }
        }

        return variantId
    }
}
